### PR TITLE
Replace defer unlocks with explicit ones

### PR DESCRIPTION
**Pull request by redbaron**
Apparently defer is not free: https://github.com/golang/go/issues/14939

### CODE REVIEW

**Comment by fabxc** [comment]
Defer is certainly not free. I agree on several of the simple replacements but especially in those functions with multiple return points, defer is safer to use.
Could you elaborate why exactly these locks (we are using defer in a ton more places)? Could you measure any performance improvements with this change?

**Comment by redbaron** [comment]
No particular reason to choosing these locks, just changed at part of the code I happen to look at

**Comment by redbaron** [comment]
go question: is it  required to unlock before panic() call?

**Comment by redbaron** [comment]
reverted to defer in multi-return funcs

**Comment by fabxc** [comment]
`panic()` just crashes the application so no need to unlock anything. In this use case `defer` is generally interesting to finish of storage writes or similar so we don't leave anything corrupted after crashing.

**Comment by fabxc** [comment]
@beorn7 your call.

I generally agree with all the un-defering done here. All these methods have a small scope and this could accumulate to a decent improvement given they are called for millions of chunks.

**Comment by beorn7** [comment]
The linked issue sounds like the defer slowness will be solved in Go1.8.

I'd be fine to do the "manual" defer in proven hot spots of the code, but not in general. Also, every occurrence should be marked is "this is done for performance reasons and should be reverted to idiomatic use once https://github.com/golang/go/issues/14939 is fixed".

I could imagine that pin and unpin are a very hot spot, but I'm not sure for every occurrence in this PR.

WRT the unlock in case of panic: People start to use the storage package independently. They might want to recover from a panic (at least for test purposes), so not unlocking in case of a panic is only acceptable if the panic only happens due to an internal programming error.

WRT naked returns: Does that have a performance impact, too? (I would hope not.) Our style guide is very restrictive about naked returns. They should in general be avoided.

**Comment by redbaron** [comment]
updated: no naked returns

**Comment by grobie** [comment]
Do we have benchmarks covering this area which show an improvement now?

**Comment by beorn7** [comment]
It would be easy to look at a CPU profiling output to see the percentage spent in any of these functions.

For pin and unpin, I would believe a relevant impact even without looking. (Those are called thousands of times per second.) Also, those functions are super short and easy. No panic/unlock problem, no problem with the unlock being "below the fold". No additional variable assignment.

I'd like to see a provable effect for the others, especially given that the problem will go away with Go1.8, most likely, and we have to revert the changes then.

**Comment by beorn7** [comment]
I finally indulged into a bit of profiling.

WRT pin/unpin: I picked our busiest server in terms of pinning (performing whopping 130k pins/sec!). Still, the profiling resulted in only 0.66% in unpin and 0.14% in pin.

I checked all the other modified functions, too,  The only one that matters (more than 1% of CPU time spent) at all is `seriesMap.get` (which has a huge share of CPU time, around 25%). I'm not sure how much of that time is just spent because of lock contention. But it's the only code where the `defer` hack at least has the potential of making a dent.

(I could imagine some of the simpler functions don't show up so much in the profiling because they were inlined by the compiler, but that would also unroll the `defer`, so it would solve our problem as a byproduct.)

Conclusion: Since this is anyway only a temporary workaround, could you just change `seriesMap.get` and add a comment that this is just a work-around until the `defer` slowness is solved (best to add a link to the bug).

Thanks.

File: storage/local/series.go [hunk 1]
```diff
@@ -57,38 +57,37 @@ func newSeriesMap() *seriesMap {
...
 
 // get returns a memorySeries for a fingerprint. Return values have the same
 // semantics as the native Go map.
 func (sm *seriesMap) get(fp model.Fingerprint) (s *memorySeries, ok bool) {
 	sm.mtx.RLock()
-	defer sm.mtx.RUnlock()
-
 	s, ok = sm.m[fp]
-	return
+	sm.mtx.RUnlock()
```
Comment (beorn7) [inline storage/local/series.go]: This is the only time where the hack matters, according to profiling.

**Comment by redbaron** [comment]
doesn't worth  it. @beorn7, out of curiosity, would you be able to publish cpu profile from that busy server?

**Comment by beorn7** [comment]
For `seriesMap.get`, it might well be worth it. To shortcut the change, I have created https://github.com/prometheus/prometheus/pull/2021.

**Comment by beorn7** [comment]
Output of `top100 -cum local.*(seriesMap|pin|refCount|isEvicted)`:

```
25.35s of 73.11s total (34.67%)
Dropped 49 nodes (cum <= 0.37s)
      flat  flat%   sum%        cum   cum%
         0     0%     0%     25.99s 35.55%  runtime.goexit
         0     0%     0%     25.46s 34.82%  github.com/prometheus/prometheus/promql.(*Engine).exec
         0     0%     0%     25.46s 34.82%  github.com/prometheus/prometheus/promql.(*Engine).execEvalStmt
         0     0%     0%     25.46s 34.82%  github.com/prometheus/prometheus/promql.(*query).Exec
         0     0%     0%     25.46s 34.82%  github.com/prometheus/prometheus/promql.Inspect
         0     0%     0%     25.46s 34.82%  github.com/prometheus/prometheus/promql.Walk
         0     0%     0%     25.46s 34.82%  github.com/prometheus/prometheus/promql.inspector.Visit
         0     0%     0%     25.46s 34.82%  github.com/prometheus/prometheus/rules.(*Group).eval.func1
         0     0%     0%     25.45s 34.81%  github.com/prometheus/prometheus/rules.(*RecordingRule).eval
         0     0%     0%     25.45s 34.81%  github.com/prometheus/prometheus/rules.RecordingRule.eval
         0     0%     0%     24.92s 34.09%  github.com/prometheus/prometheus/promql.(*Engine).populateIterators
         0     0%     0%     24.92s 34.09%  github.com/prometheus/prometheus/promql.(*Engine).populateIterators.func1
         0     0%     0%     24.91s 34.07%  github.com/prometheus/prometheus/storage/local.(*MemorySeriesStorage).QueryRange
     6.83s  9.34%  9.34%     24.83s 33.96%  github.com/prometheus/prometheus/storage/local.(*seriesMap).get
         0     0%  9.34%     24.73s 33.83%  github.com/prometheus/prometheus/storage/local.(*MemorySeriesStorage).metricForRange
         0     0%  9.34%     24.20s 33.10%  github.com/prometheus/prometheus/storage/local.(*MemorySeriesStorage).metricsForLabelMatchers
     3.79s  5.18% 14.53%      7.39s 10.11%  sync.(*RWMutex).RUnlock
     6.95s  9.51% 24.03%      6.95s  9.51%  sync/atomic.AddUint32
     4.14s  5.66% 29.69%      4.35s  5.95%  runtime.mapaccess2_fast64
     0.13s  0.18% 29.87%      3.60s  4.92%  sync.(*RWMutex).RLock
         0     0% 29.87%      2.21s  3.02%  runtime.systemstack
     0.26s  0.36% 30.23%      1.62s  2.22%  runtime.deferproc
     0.36s  0.49% 30.72%      1.37s  1.87%  runtime.deferreturn
     0.21s  0.29% 31.01%      1.02s  1.40%  runtime.deferproc.func1
         0     0% 31.01%      0.72s  0.98%  github.com/prometheus/prometheus/storage/local.(*MemorySeriesStorage).preloadChunksForRange
     0.06s 0.082% 31.09%      0.65s  0.89%  runtime.writebarrierptr_nostore1.func1
     0.54s  0.74% 31.83%      0.63s  0.86%  runtime.newdefer
     0.14s  0.19% 32.02%      0.59s  0.81%  runtime.gcmarkwb_m
     0.35s  0.48% 32.50%      0.55s  0.75%  github.com/prometheus/prometheus/storage/local.(*chunkDesc).unpin
         0     0% 32.50%      0.54s  0.74%  github.com/prometheus/prometheus/promql.(*Engine).closeIterators
         0     0% 32.50%      0.54s  0.74%  github.com/prometheus/prometheus/promql.(*Engine).closeIterators.func1
         0     0% 32.50%      0.54s  0.74%  github.com/prometheus/prometheus/storage/local.(*boundedIterator).Close
         0     0% 32.50%      0.54s  0.74%  github.com/prometheus/prometheus/storage/local.(*memorySeriesIterator).Close
         0     0% 32.50%      0.53s  0.72%  github.com/prometheus/prometheus/storage/local.(*MemorySeriesStorage).seriesForRange
     0.05s 0.068% 32.57%      0.51s   0.7%  runtime.deferreturn.func1
     0.40s  0.55% 33.11%      0.46s  0.63%  runtime.freedefer
     0.08s  0.11% 33.22%      0.46s  0.63%  runtime.writebarrierptr
     0.16s  0.22% 33.44%      0.45s  0.62%  runtime.shade
     0.19s  0.26% 33.70%      0.38s  0.52%  runtime.writebarrierptr_nostore1
     0.02s 0.027% 33.73%      0.34s  0.47%  github.com/prometheus/prometheus/storage/local.(*seriesMap).iter.func1
     0.23s  0.31% 34.04%      0.23s  0.31%  runtime.heapBitsForObject
     0.21s  0.29% 34.33%      0.21s  0.29%  runtime.aeshash64
         0     0% 34.33%      0.19s  0.26%  github.com/prometheus/prometheus/storage/local.(*memorySeries).preloadChunks
         0     0% 34.33%      0.19s  0.26%  github.com/prometheus/prometheus/storage/local.(*memorySeries).preloadChunksForRange
         0     0% 34.33%      0.17s  0.23%  github.com/prometheus/prometheus/retrieval.(*ruleLabelsAppender).Append
         0     0% 34.33%      0.17s  0.23%  github.com/prometheus/prometheus/retrieval.(*scrapeLoop).append
         0     0% 34.33%      0.17s  0.23%  github.com/prometheus/prometheus/retrieval.(*scrapeLoop).run
         0     0% 34.33%      0.17s  0.23%  github.com/prometheus/prometheus/retrieval.ruleLabelsAppender.Append
         0     0% 34.33%      0.17s  0.23%  github.com/prometheus/prometheus/storage.(*Fanout).Append
         0     0% 34.33%      0.17s  0.23%  github.com/prometheus/prometheus/storage.Fanout.Append
         0     0% 34.33%      0.17s  0.23%  github.com/prometheus/prometheus/storage/local.(*MemorySeriesStorage).Append
     0.02s 0.027% 34.36%      0.16s  0.22%  runtime.chansend1
     0.15s  0.21% 34.56%      0.15s  0.21%  runtime.memmove
         0     0% 34.56%      0.15s  0.21%  runtime.semacquire
         0     0% 34.56%      0.15s  0.21%  sync.runtime_Semacquire
     0.04s 0.055% 34.62%      0.14s  0.19%  runtime.chansend
     0.04s 0.055% 34.67%      0.14s  0.19%  runtime.lock
```

Not sure if the raw profile contains any internal information I'm not allowed to share.

**Pull request closed without merging**